server/src/main/java/org/opensearch/gateway/AsyncShardBatchFetch.java

```diff
@@ -0,0 +1,341 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.gateway;
+
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.support.nodes.BaseNodeResponse;
+import org.opensearch.action.support.nodes.BaseNodesResponse;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.cluster.node.DiscoveryNodes;
+import org.opensearch.common.logging.Loggers;
+import org.opensearch.core.index.shard.ShardId;
+import org.opensearch.indices.store.ShardAttributes;
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * Implementation of AsyncShardFetch with batching support. This class is responsible for executing the fetch
+ * part using the base class {@link AsyncShardFetch}. Other functionalities needed for a batch are only written here.
+ * Cleanup of failed shards is necessary in a batch and based on that a reroute should be triggered to take care of
+ * those in the next run. This separation also takes care of the extra generic type V which is only needed for batch
+ * transport actions like {@link TransportNodesListGatewayStartedShardsBatch}.
+ *
+ * @param <T> Response type of the transport action.
+ * @param <V> Data type of shard level response.
+ *
+ * @opensearch.internal
+ */
+public abstract class AsyncShardBatchFetch<T extends BaseNodeResponse, V> extends AsyncShardFetch<T> {
+
+    private final Consumer<ShardId> removeShardFromBatch;
+    private final List<ShardId> failedShards;
+
+    @SuppressWarnings("unchecked")
+    AsyncShardBatchFetch(
+        Logger logger,
+        String type,
+        Map<ShardId, ShardAttributes> shardAttributesMap,
+        AsyncShardFetch.Lister<? extends BaseNodesResponse<T>, T> action,
+        String batchId,
+        Class<V> clazz,
+        BiFunction<DiscoveryNode, Map<ShardId, V>, T> responseGetter,
+        Function<T, Map<ShardId, V>> shardsBatchDataGetter,
+        Supplier<V> emptyResponseBuilder,
+        Consumer<ShardId> failedShardHandler,
+        Function<V, Exception> getResponseException,
```

Comment:
this looks like really long list of arguments

Reply:
Yes, all these constructor or getter are needed to keep the class generic for both primary and replica transport response types. 

We can move them to a separate inner class like `BatchCacheTransformer`, but we'll still need it here to create the object of that new inner class.

Reply:
you can define an interface In AsyncShardBatchFetch

    interface ShardCacheEntry<T> {
        boolean isEmpty();
        T getObject();
        T getEmpty();
    }

would it simplify your list of args?

Reply:
Also with this interface do you still need NodeEntry<V> class?

Reply:
`NodeEntry` = This class defines the cache entry for a particular node's response. Actual data is saved in an array format. 
NodeEntry class is different [for single shard](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/gateway/AsyncShardFetch.java#L321-L339) and batch shard case, so it'll still be needed irrespective of the interface existence. 

On the interface side 

**Writing current argument names of ShardBatchCache class in []**

**boolean isEmpty** - needs to be returned by Shard Level Response [**isEmpty**]
**V getEmpty** - needs to be returned by Shard Level Response class [**emptyResponseBuilder**]

**T getObject** - needs to be returned by Node Level Response class [**responseConstructor**]
**Map<ShardId, V> getShardsData** - needs to be returned by Node level response class [**shardsBatchDataGetter**]

I don’t think all of these can be in single interface.

Reply:
I'm thinking of adding one Factory class to reduce 2 arguments to 1 . Based on primary or replica, it'll be able to create new responses and return relevant shard batch data.

Let me know if it looks okay ?

```
public class ShardBatchResponseFactory<T extends BaseNodeResponse, V> {
    private final boolean primary;

    public ShardBatchResponseFactory(boolean primary) {
        this.primary = primary;
    }

    public T getNewResponse(DiscoveryNode node, Map<ShardId, V> shardData) {
        if (primary) {
            return (T) new NodeGatewayStartedShardsBatch(node, (Map<ShardId,
                GatewayStartedShard>) shardData);
        } else {
            return (T) new NodeStoreFilesMetadataBatch(node, (Map<ShardId,
                NodeStoreFilesMetadata>) shardData);
        }
    }

    public Map<ShardId, V> getShardBatchData(T response) {
        if (primary) {
            return (Map<ShardId, V>) ((NodeGatewayStartedShardsBatch)response).getNodeGatewayStartedShardsBatch();
        } else {
            return (Map<ShardId, V>) ((NodeStoreFilesMetadataBatch)response).getNodeStoreFilesMetadataBatch();
        }
    }

}
```


This will change the constructor to look like 

```
public ShardBatchCache(
            Logger logger,
            String type,
            Map<ShardId, ShardAttributes> shardAttributesMap,
            String logKey,
            Class<V> clazz,
            V emptyResponse,
            Function<V, Boolean> isEmptyResponse,
            ShardBatchResponseFactory<T, V> responseFactory
        )
```